File: src/main/java/lib/math/controllers/pathfollowing/ModuleForcesPathFollower.java
package lib.math.controllers.pathfollowing;

import choreo.trajectory.SwerveSample;
import edu.wpi.first.math.VecBuilder;
import edu.wpi.first.math.Vector;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.numbers.N2;
import frc.robot.subsystems.drive.Drive;
import lib.math.controllers.gains.PIDGains;
import org.littletonrobotics.junction.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiConsumer;
import java.util.function.Supplier;

/**
 * Path follower that additionally returns desired module forces for force-based
 * control.
 */
public final class ModuleForcesPathFollower implements PathFollower {
  private final Drive drive;
  private final PIDController xPID;
  private final PIDController yPID;
  private final PIDController thetaPID;
  private final BiConsumer<ChassisSpeeds, List<Vector<N2>>> speedConsumer;
  private final Supplier<Pose2d> poseSupplier;

  public ModuleForcesPathFollower(
      Drive drive,
      PIDGains xPidGains,
      PIDGains yPidGains,
      PIDGains thetaPidGains,
      BiConsumer<ChassisSpeeds, List<Vector<N2>>> speedConsumer,
      Supplier<Pose2d> poseSupplier) {
    this.drive = drive;
    this.xPID = new PIDController(xPidGains.getKP(), xPidGains.getKI(), xPidGains.getKD());
    this.yPID = new PIDController(yPidGains.getKP(), yPidGains.getKI(), yPidGains.getKD());
    this.thetaPID = new PIDController(thetaPidGains.getKP(), thetaPidGains.getKI(), thetaPidGains.getKD());
    this.thetaPID.enableContinuousInput(-Math.PI, Math.PI);
    this.speedConsumer = speedConsumer;
    this.poseSupplier = poseSupplier;
  }

  @Override
  public void accept(SwerveSample sample) {
    Pose2d pose = poseSupplier.get();

    double xOutput = xPID.calculate(pose.getX(), sample.x);
    double yOutput = yPID.calculate(pose.getY(), sample.y);
    double thetaOutput = thetaPID.calculate(pose.getRotation().getRadians(), sample.heading);

    Logger.recordOutput("drivebase/auto/xError", sample.x - pose.getX());
    Logger.recordOutput("drivebase/auto/yError", sample.y - pose.getY());
    Logger.recordOutput("drivebase/auto/thetaError", sample.heading - pose.getRotation().getRadians());

    Logger.recordOutput("drivebase/auto/samplePose", Pose2d.struct, sample.getPose());
    Logger.recordOutput("drivebase/auto/pose", Pose2d.struct, pose);

    double[] moduleForcesX = sample.moduleForcesX();
    double[] moduleForcesY = sample.moduleForcesY();
    int moduleCount = Math.min(moduleForcesX.length, moduleForcesY.length);
    List<Vector<N2>> moduleForces = new ArrayList<>(moduleCount);
    for (int i = 0; i < moduleCount; i++) {
      moduleForces.add(VecBuilder.fill(moduleForcesX[i], moduleForcesY[i]));
    }

    speedConsumer.accept(
        ChassisSpeeds.fromFieldRelativeSpeeds(
            xOutput + sample.vx,
            yOutput + sample.vy,
            thetaOutput + sample.omega,
            pose.getRotation()),
        moduleForces);
  }
}

File: src/main/java/lib/controllers/pathfollowing/ModuleForcesPathFollower.kt
package lib.controllers.pathfollowing

import choreo.trajectory.SwerveSample
import edu.wpi.first.math.VecBuilder
import edu.wpi.first.math.Vector
import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.math.geometry.Pose2d
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.math.numbers.N2
import frc.robot.subsystems.swerve.Drivebase
import lib.controllers.gains.PIDGains
import org.littletonrobotics.junction.Logger
import java.util.function.BiConsumer
import java.util.function.Consumer
import java.util.function.Supplier
import kotlin.math.PI

class ModuleForcesPathFollower(
    private val drivebase: Drivebase,
    xPidGains: PIDGains,
    yPidGains: PIDGains,
    thetaPidGains: PIDGains,
    private val speedConsumer: BiConsumer<ChassisSpeeds, List<Vector<N2>>>,
    private val poseSupplier: Supplier<Pose2d>
): PathFollower {

    private val xPID = PIDController(xPidGains.kP, xPidGains.kI, xPidGains.kD)
    private val yPID = PIDController(yPidGains.kP, yPidGains.kI, yPidGains.kD)
    private val thetaPID = PIDController(thetaPidGains.kP, thetaPidGains.kI, thetaPidGains.kD)
        .apply { enableContinuousInput(-PI, PI) }

    override fun accept(sample: SwerveSample) {
        val pose = poseSupplier.get()

        val xOutput = xPID.calculate(pose.translation.x, sample.x)
        val yOutput = yPID.calculate(pose.translation.y, sample.y)
        val thetaOutput = thetaPID.calculate(pose.rotation.radians, sample.heading)

        Logger.recordOutput("drivebase/auto/xError", xPID.error)
        Logger.recordOutput("drivebase/auto/yError", yPID.error)
        Logger.recordOutput("drivebase/auto/thetaError", thetaPID.error)

        Logger.recordOutput("drivebase/auto/samplePose", Pose2d.struct, sample.pose)
        Logger.recordOutput("drivebase/auto/pose", Pose2d.struct, pose)

        speedConsumer.accept(
            ChassisSpeeds.fromFieldRelativeSpeeds(
                xOutput + sample.vx,
                yOutput + sample.vy,
                thetaOutput + sample.omega,
                pose.rotation
            ),
            sample.moduleForcesX().zip(sample.moduleForcesY()).map { VecBuilder.fill(it.first, it.second) }
        )
    }
}